browse/services/document/metadata.py

```diff
@@ -325,6 +329,97 @@ def get_previous_id(self, identifier: Identifier):
+        return None
+
+    def get_dissemination_formats(self,
+                                  docmeta: DocMetadata,
+                                  format_pref: str = None,
+                                  add_sciencewise: bool = False) -> List[str]:
+        """
+        Get a list of formats that can be disseminated for this DocMetadata.
+
+        Several checks are performed to determine available dissemination
+        formats:
+            1. a check for source files with specific, valid file name
+               extensions (i.e. for a subset of the allowed source file name
+               extensions, the dissemintation formats are predictable)
+            2. if formats cannot be inferred from source file, inspect the
+               source type in the document metadata.
+
+        Format names are strings. These include 'src', 'pdf', 'ps', 'html',
+        'pdfonly', 'other', 'dvi', 'ps(400)', 'ps(600)', 'nops'.
+
+        Parameters
+        ----------
+        docmeta : :class:`DocMetadata`
+        format_pref : str
+            The format preference string.
+        add_sciencewise : bool
+            Specify whether to include 'sciencewise_pdf' format in list.
+
+        Returns
+        -------
+        List[str]
+            A list of format strings.
+
+        """
+        formats = []
+
+        # first, get possible list of formats based on available source file
+        source_file_path = self._get_source_path(docmeta)
+        source_file_formats = formats_from_source_file_name(source_file_path)
+        if source_file_formats:
+            formats.extend(source_file_formats)
+        else:
+            # check source type from metadata, with consideration of
+            # user format preference and cache
+            version = docmeta.version
+            format_code = docmeta.version_history[version - 1].source_type.code
+            cached_ps_file_path = cache.get_cache_file_path(
+                docmeta,
+                'ps')
+            cache_flag = False
+            if cached_ps_file_path \
+                    and os.path.getsize(cached_ps_file_path) == 0 \
+                    and source_file_path \
+                    and os.path.getmtime(source_file_path) \
+                    < os.path.getmtime(cached_ps_file_path):
+                cache_flag = True
+            source_type_formats = formats_from_source_type(format_code,
+                                                           format_pref,
+                                                           cache_flag)
+            if source_type_formats:
+                formats.extend(source_type_formats)
+
+        # Separate check for ScienceWISE annotated PDF
+        if add_sciencewise:
+            if formats and formats[-1] == 'other':
+                formats.insert(-1, 'sciencewise_pdf')
+            else:
+                formats.append('sciencewise_pdf')
+
+        return formats
+
+    def get_ancillary_files(self, docmeta: DocMetadata) \
```

Comment:
Is there logic here for why an `Option[List[_]]` is needed? Generally, it is ok to return an empty List in place of None, but there could be exceptions if we should really treat an empty list differently from `None` in the calling code.

This comment applies to several other places in the code as well.

Reply:
There was really no need to return `None` in all of the cases I had used `Optional[List[_]]` so I updated the functions to return `List[_]`.